experimental/ast/type.go

```diff
@@ -42,6 +43,23 @@ type TypeKind int8
 // This type is used in lieu of a putative Type interface type to avoid heap
 // allocations in functions that would return one of many different Type*
 // types.
+//
+// # Grammar
+//
+//	Type := TypePath | TypePrefixed | TypeGeneric
+//
+// Note that there is an ambiguity when parsing a Type followed by a Path:
+// "optional optional foo" could be parsed as:
```

Comment:
This isn't quite clear. I think the root issue is whether parsing a Type should be greedy, in which case it consumes all three tokens (first example) or not (second example).  This only is an issue in practice in the grammar when the Type is followed by a Path.

So maybe the last paragraph can state that parsing is always greedy _except_ when we know a Path can follow, in which case we don't try to consume the last/trailing path.